fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java

```diff
@@ -144,6 +144,69 @@ public ImmutableMap<Long, ComputeNode> getIdComputeNode() {
         }
     }
 
+    /**
+     * The schedule policy is suitable in shared nothing mode.
+     */
+    @Test
+    public void testSelectBackendAndComputeNode() {
+        new MockUp<SystemInfoService>() {
+            @Mock
+            public ImmutableMap<Long, ComputeNode> getIdToBackend() {
+                return availableId2Backend;
+            }
+
+            @Mock
+            public ImmutableMap<Long, ComputeNode> getIdComputeNode() {
+                return availableId2ComputeNode;
+            }
+        };
+
+        DefaultWorkerProvider.Factory workerProviderFactory = new DefaultWorkerProvider.Factory();
+        DefaultWorkerProvider workerProvider;
+        List<Integer> numUsedComputeNodesList = ImmutableList.of(-1, 0, 2, 4);
+
+        // test ComputeNode only
+        for (Integer numUsedComputeNodes : numUsedComputeNodesList) {
+            workerProvider =
+                    workerProviderFactory.captureAvailableWorkers(GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo(),
+                            true, numUsedComputeNodes, "compute_nodes_only", WarehouseManager.DEFAULT_WAREHOUSE_ID);
+            List<Long> selectedWorkerIdsList = workerProvider.getAllAvailableNodes();
+            for (Long selectedWorkerId : selectedWorkerIdsList) {
+                Assert.assertTrue("selectedWorkerId:" + selectedWorkerId,
+                        availableId2ComputeNode.containsKey(selectedWorkerId));
+            }
+        }
+        // test Backend only
+        for (Integer numUsedComputeNodes : numUsedComputeNodesList) {
+            workerProvider =
+                    workerProviderFactory.captureAvailableWorkers(GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo(),
+                            false, numUsedComputeNodes, "compute_nodes_only", WarehouseManager.DEFAULT_WAREHOUSE_ID);
+            List<Long> selectedWorkerIdsList = workerProvider.getAllAvailableNodes();
+            Assert.assertEquals(availableId2Backend.size(), selectedWorkerIdsList.size());
+            for (Long selectedWorkerId : selectedWorkerIdsList) {
+                Assert.assertTrue("selectedWorkerId:" + selectedWorkerId,
+                        availableId2Backend.containsKey(selectedWorkerId));
+            }
+        }
+        // test Backend and ComputeNode
+        for (Integer numUsedComputeNodes : numUsedComputeNodesList) {
```

Comment:
when numUsedComputeNode < ComputeNodes.size(), and it is the `all_nodes` policy, what's the expected behavior?

when it is `compute_nodes_only`, the candidate workers has only `numUsedComputeNode` compute nodes, when it is `all_nodes`, the candidate workers will has `numUsedComputeNode` compute nodes plus all the be nodes based on this PR's implementation, looks like not make sense?

Reply:
if the user understand the `preferComputeNode/useComputeNodes/computation_fragment_scheduling_policy` three session variable 's correlation and cooperation , and the useComputeNodes is set by user self，will it make sense when the case of numUsedComputeNode < ComputeNodes.size() ?

Reply:
maybe just don't want the query occupy too much resource? e.g. for some offline batch jobs, limit the number of used compute nodes to ETL data into sr local table.

Reply:
i think make clear of `preferComputeNode/useComputeNodes/computation_fragment_scheduling_policy` three session variable 's correlation and cooperation in the user document，when the case like you say，user would understand the `all_nodes=numUsedComputeNode cn plus all be`

Reply:
it is the scope of `computation_fragment_scheduling_policy` session variable, will it just control if take BE as CN this single point, or will it be also to overwrite the `useComputeNodes` settings?

I am opt to the former. this PR's implementation is the later, and not fully overwritten, just partially overwritten.

Reply:
i will update the pr accoding to the former. the former is really clear more than later, just think about for a moment.

Reply:
@kevincai i updated the pr to adjust the scope of computation_fragment_scheduling_policy, could you please approval the workflow ?